### PR TITLE
[Debug] Different trace message style on application code

### PR DESCRIPTION
**Pull request by rentalhost**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #23144
| License       | MIT
| Doc PR        | no

With this PR, the debug stacktrace now differ the background-color of the row when the file is outside the vendor folder, make more clear to identify where the problem starts (that is generally from user code).

**Example:**

image

### CODE REVIEW

**Comment by javiereguiluz** [comment]
@rentalhost by default we expand the first code sample outside the vendor/ dir, so you can go directly to that (because, as you said, there's nothing to do in vendor/ because you can't change that).

I see that you are using Laravel ... maybe that feature is not working in your case? By the way, you won't see this in the next Laravel version to be released very soon, because they are adding back the [Whops library](https://filp.github.io/whoops/).

**Comment by rentalhost** [comment]
@javiereguiluz Well, for some reason, I see all mixed on stacktrace: vendor and user code. It make a lot of sense to me, once that the user code calls some vendor method that, via Closures, for instance, call an user code again (and so on). But, I still have a big difficulty that is to locate "what is my code and what is not" (in general, is on my code that I should starts to debug to find out a problem).

This picture is from a code example from Laravel (I injected this modified class to my project to test it before PR), but I tried too on Tests and it works fine (on case, I could identify easily what is user and vendor code).

And yep, I know that Laravel will back to Whoops, but for some reason seems that symfony/debug is being used on current versions. Possibly I will not see this change on Laravel on future, but I guess that is a useful feature for symfony users anyway.

File: src/Symfony/Component/Debug/ExceptionHandler.php [hunk 1]
```diff
@@ -411,4 +414,27 @@ private function getSymfonyGhostAsSvg()
...
+    private static function getTraceClassname($trace)
+    {
+        if (empty($trace['file'])) {
+            return null;
+        }
+
+        $vendorRoot = static::getVendorRoot();
+        $traceFileRelative = str_replace($vendorRoot, '', realpath($trace['file']));
+
+        if (strpos($traceFileRelative, '/vendor/') === 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: the generic logic to get vendors and composer root is a bit more involving, see `LinkStub::getComposerRoot()`:
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/VarDumper/Caster/LinkStub.php#L64

**Comment by javiereguiluz** [comment]
I don't disagree with this feature ... but I disagree with the proposed visual design of this feature. If this feature is finally accepted, I'll propose a different design to see if it's accepted by the reviewers. Thanks!

File: src/Symfony/Component/Debug/ExceptionHandler.php (hunk 1 above)
Comment (rentalhost) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: It is a great idea reuse that. But currently the VarDumper is not a dependency of Debug component. Can I turn it a dependency?

**Comment by rentalhost** [comment]
@javiereguiluz What you like to change? I can do it for now, once that it is not the "final" yet. We can discuss about your idea here.

File: src/Symfony/Component/Debug/ExceptionHandler.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: just copy/paste :)

**Comment by greg0ire** [comment]
It would be great to have similar changes for the CLI output, and for the plain text stack trace (obviously not with color in that particular case).

**Comment by fabpot** [comment]
@javiereguiluz I let you drive this PR forward :)

File: src/Symfony/Component/Debug/ExceptionHandler.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: ping @rentalhost, ok for this comment?

Comment (rentalhost) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: It's ok, I will cp. Thanks!

Comment (rentalhost) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: Hm... How much this method is better than the currently implemented? I mean, I have based the vendor directory on the `ClassLoader` class from composer, and it is more simple than `LinkStub::getComposerRoot()` method. And just copy/paste don't solve the problem, I should adapt the code anyway.

Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: @rentalhost sorry for the delay. The code to look for vendors in LinkStub handles two things yours doesn't:
- there can be several vendor directories - eg the global one, the app's, phpunit's when running test, to consider a not uncommon situation
- the vendor directories can have a [different name than "vendor"](https://getcomposer.org/doc/06-config.md#vendor-dir)

**Comment by nicolas-grekas** [comment]
@rentalhost any time to finish this PR by the end of the week? feature freeze is coming :)

**Comment by javiereguiluz** [comment]
I'd like to not rush this feature until we can think of a better visual design. Thanks!

**Comment by nicolas-grekas** [comment]
Moving to 4.1 then.

**Comment by fabpot** [comment]
@javiereguiluz What do we do here? There is no rush, but after 6 months of wait, I think we need to take a decision here.

**Comment by fabpot** [comment]
Closing in favor of #26671

Thanks @rentalhost for your work on this pull request. It made Symfony better in the end even if your code is not going to be merged.

**Pull request closed without merging**